agent/setup.go

```diff
@@ -55,11 +55,18 @@ type BaseDeps struct {
 
 type ConfigLoader func(source config.Source) (config.LoadResult, error)
 
-func NewBaseDeps(configLoader ConfigLoader, logOut io.Writer, providedLogger hclog.InterceptLogger) (BaseDeps, error) {
+func NewBaseDeps(configLoader ConfigLoader, logOut io.Writer, providedLogger hclog.InterceptLogger) (BaseDeps, func(), error) {
+	var balancerBuilder *balancer.Builder
+	cleanup := func() {
+		if balancerBuilder != nil {
+			balancerBuilder.Deregister()
+		}
+	}
```

Comment:
This feels a bit gross. But so is mutating global state here! We probably need to rethink `BaseDeps` in general.

Reply:
I can think of the other alternatives:
 1. Pass in a context that lives for the live of the agent. I think we already have ones created from stopChans in a few other places and just run deregistration when the context is closed.
 2. Since `BaseDeps` is a type, make it an `io.Closer` and have the agent call `Close` during shutdown, then hook deregistration etc. up to it here before returning.

Got to be honest though, I don't think either of them is clearly better than what you've done here so I'd be inclined to keep it this way unless/until this becomes really unpleasant to maintain for some reason 🤷 .